src/wallet/wallet.cpp

```diff
@@ -93,72 +94,116 @@ CPubKey CWallet::GenerateNewKey()
+    if (GetKey(hdChain.masterKeyID, key)) {
+        CExtKey masterKey;             //hd master key
+        CExtKey accountKey;            //key at m/0'
+        CExtKey externalChainChildKey; //key at m/0'/0'
+        CExtKey childKey;              //key at m/0'/0'/<n>'
+
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        // use hardened derivation (child keys >= 0x80000000 are hardened after bip32)
+        masterKey.Derive(accountKey, BIP32_HARDENED_KEY_LIMIT);
+
+        // derive m/0'/0'
+        accountKey.Derive(externalChainChildKey, BIP32_HARDENED_KEY_LIMIT);
+
+        // derive child key at next index, skip keys already known to the wallet
+        do {
+            // always derive hardened keys
+            // childIndex | BIP32_HARDENED_KEY_LIMIT = derive childIndex in hardened child-index-range
+            // example: 1 | BIP32_HARDENED_KEY_LIMIT == 0x80000001 == 2147483649
+            externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | BIP32_HARDENED_KEY_LIMIT);
+            metadata.hdKeypath = "m/0'/0'/" + std::to_string(hdChain.nExternalChainCounter) + "'";
+            metadata.hdMasterKeyID = hdChain.masterKeyID;
+            // increment childkey index
+            hdChain.nExternalChainCounter++;
+        } while (HaveKey(childKey.key.GetPubKey().GetID()));
+        secret = childKey.key;
+        pubKey = childKey.key.GetPubKey();
+        // update the chain model in the database
+        if (!CWalletDB(strWalletFile).WriteHDChain(hdChain))
+            throw std::runtime_error(std::string(__func__) + ": Writing HD chain model failed");
+        return true;
+    }
+    else if (hdChain.isWatchOnly) {
+        CExtPubKey& masterKey = hdChain.masterPubKey;             //hd master key
+        CExtPubKey accountKey;            //key at m/0
+        CExtPubKey externalChainChildKey; //key at m/0/0
+        CExtPubKey childKey;              //key at m/0/0/<n>
+
+        // derive m/0
+        masterKey.Derive(accountKey, BIP32_NONHARDENED_KEY_LIMIT);
+
+        // derive m/0/0
+        accountKey.Derive(externalChainChildKey, BIP32_NONHARDENED_KEY_LIMIT);
+
+        // derive child key at next index, skip keys already known to the wallet
+        do {
+            externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | BIP32_NONHARDENED_KEY_LIMIT);
+            metadata.hdKeypath = "m/0/0/" + std::to_string(hdChain.nExternalChainCounter) + "'";
+            metadata.hdMasterKeyID = hdChain.masterKeyID;
+            // increment childkey index
+            hdChain.nExternalChainCounter++;
+        } while (HaveWatchOnly(GetScriptForDestination(childKey.pubkey.GetID())));
+        pubKey = childKey.pubkey;
+        // update the chain model in the database
+        if (!CWalletDB(strWalletFile).WriteHDChain(hdChain))
+            throw std::runtime_error(std::string(__func__) + ": Writing HD chain model failed");
+        return false;
```

Comment:
Yes. I once did a PR where we separate the key/pubkey records: https://github.com/bitcoin/bitcoin/pull/9298
I guess this would be the cleaner solution... but maybe later.